inline snapshot does not resolve NewType types

The following is the result of a pytest run which should create inline snapshots (`uv run python -m pytest --inline-snapshot=create  tests/test_dummy.py`):

```py
from inline_snapshot import snapshot as inline_snapshot
from typing import NewType
from pydantic import BaseModel

SomeID = NewType("SomeID", int)


class Something(BaseModel):
    some_id: SomeID


def test_whatever() -> None:
    """Test the effect of inline snapshot with NewType."""
    a = Something(some_id=SomeID(1))
    assert a == inline_snapshot(Something(some_id=1))
``` 

Unfortunately, it fails type checking, at least in some type checkers: intelij marks it yellow

<img width="1062" height="228" alt="Image" src="https://github.com/user-attachments/assets/5b01562c-eed2-4c4b-930a-904c5eaf9218" />

ty errors:

```text
uvx  ty check tests/test_dummy.py
error[invalid-argument-type]: Argument is incorrect
  --> tests/test_dummy.py:15:43
   |
13 |     """Test the effect of inline snapshot with NewType."""
14 |     a = Something(some_id=SomeID(1))
15 |     assert a == inline_snapshot(Something(some_id=1))
   |                                           ^^^^^^^^^ Expected `SomeID`, found `Literal[1]`
   |
info: rule `invalid-argument-type` is enabled by default
```

mypy without a pydantic plugin (with plugin there is no error):

```text
mypy tests/test_dummy.py        
tests/test_dummy.py:15: error: Argument "some_id" to "Something" has incompatible type "int"; expected "SomeID"  [arg-type]
Found 1 error in 1 file (checked 1 source file)
```
It would be amazing if inline snapshot could add the proper type when the type is a NewType one.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -15,7 +15,7 @@
 
 ## Fixed
 
-- Fixed `UsageError("unmanaged values can not be compared with snapshots")` raised during session teardown when using `-k` to filter tests ([#355](https://github.com/15r10nk/inline-snapshot/issues/355)). This was caused by inline-snapshot trying to update snapshots that were never compared. This is a rare edge case that caused problems when matchers were used, so it has been removed.
+- Fixed `UsageError("unmanaged values cannot be compared with snapshots")` raised during session teardown when using `-k` to filter tests ([#355](https://github.com/15r10nk/inline-snapshot/issues/355)). This was caused by inline-snapshot trying to update snapshots that were never compared. This is a rare edge case that caused problems when matchers were used, so it has been removed.
 
 <a id='changelog-0.32.3'></a>
 # 0.32.3 — 2026-02-24
@@ -740,7 +740,7 @@
 
 ## Removed
 
-- removed the `"Programming Language :: Python :: Implementation :: PyPy"` classifier which was incorrect, because inline-snapshot can not fix snapshots on pypy.
+- removed the `"Programming Language :: Python :: Implementation :: PyPy"` classifier which was incorrect, because inline-snapshot cannot fix snapshots on pypy.
   inline-snapshot now enforces `--inline-snapshot=disable` when used with an implementation other than cpython, which allows it to be used in packages that want to support pypy.
 
 ## Added

--- a/CONTRIBUTING.md
+++ b/CONTRIBUTING.md
@@ -19,7 +19,7 @@ The goal here is to find different edge cases which might have bugs.
 
 However, it is possible to exclude some code from the coverage.
 
-Code can be marked with `pragma: no cover`, if it can not be tested for some reason.
+Code can be marked with `pragma: no cover`, if it cannot be tested for some reason.
 This makes it easy to spot uncovered code in the source.
 
 Impossible conditions can be handled with `assert False`.

--- a/changelog.d/20260409_205903_15r10nk-git.md
+++ b/changelog.d/20260409_205903_15r10nk-git.md
@@ -1,0 +1,3 @@
+### Fixed
+
+- Pydantic models, attrs and dataclasses with `NewType`-typed fields now generate correct snapshots that wrap the field value with the `NewType` constructor (e.g. `Something(some_id=SomeID(1))` instead of `Something(some_id=1)`). This could only be fixed for Python >= 3.10.

--- a/src/inline_snapshot/_code_repr.py
+++ b/src/inline_snapshot/_code_repr.py
@@ -118,9 +118,7 @@ def value_code_repr(obj):
     if not type(obj) == type(obj):  # pragma: no cover
         # this was caused by https://github.com/samuelcolvin/dirty-equals/issues/104
         # dispatch will not work in cases like this
-        return (
-            f"HasRepr({repr(type(obj))}, '< type(obj) can not be compared with == >')"
-        )
+        return f"HasRepr({repr(type(obj))}, '< type(obj) cannot be compared with == >')"
 
     result = code_repr_dispatch(obj)
 

--- a/src/inline_snapshot/_customize/_builder.py
+++ b/src/inline_snapshot/_customize/_builder.py
@@ -35,6 +35,11 @@ class Builder:
         else:
             return v
 
+    def _get_value(self, value):
+        if isinstance(value, Custom):
+            return value._eval()
+        return value
+
     def _get_handler(self, v) -> Custom:
 
         from inline_snapshot._global_state import state
@@ -120,9 +125,9 @@ customized_representation={result!r}
         The value doesn't have to be a Custom node and is converted by inline-snapshot if needed.
         """
         if isinstance(default, Custom):
-            raise UsageError("default value can not be an Custom value")
+            raise UsageError("default value cannot be a Custom value")
 
-        if value == default:
+        if self._get_value(value) == default:
             return CustomDefault(value=self._get_handler_recursive(value))
         return value
 

--- a/src/inline_snapshot/_customize/_custom_call.py
+++ b/src/inline_snapshot/_customize/_custom_call.py
@@ -68,5 +68,5 @@ class CustomCall(Custom):
         except Exception as e:
             call_args = args + [f"{k}={v}" for k, v in kwargs.items()]
             raise TypeError(
-                f"can not call {self.function}({', '.join(map(str,call_args))})"
+                f"cannot call {self.function}({', '.join(map(str,call_args))})"
             ) from e

--- a/src/inline_snapshot/_external/_external.py
+++ b/src/inline_snapshot/_external/_external.py
@@ -149,7 +149,7 @@ class External(ExternalBase):
 
         if not location.stem:
             raise UsageError(
-                f"can not load external object from an non existing location {location.to_str()!r}"
+                f"cannot load external object from a non-existing location {location.to_str()!r}"
             )
         storage = state().all_storages[location.storage]
 

--- a/src/inline_snapshot/_external/_external_base.py
+++ b/src/inline_snapshot/_external/_external_base.py
@@ -57,12 +57,12 @@ class ExternalBase:
 
         elif isinstance(other, ExternalBase):
             raise UsageError(
-                f"you can not compare {external_type}(...) with {external_type}(...)"
+                f"you cannot compare {external_type}(...) with {external_type}(...)"
             )
 
         elif isinstance(other, GenericValue):
             raise UsageError(
-                f"you can not compare {external_type}(...) with snapshot(...)"
+                f"you cannot compare {external_type}(...) with snapshot(...)"
             )
 
         first_comparison = not self._was_compared

--- a/src/inline_snapshot/_external/_external_file.py
+++ b/src/inline_snapshot/_external/_external_file.py
@@ -40,7 +40,7 @@ class ExternalFile(ExternalBase, SnapshotRefBase):
             return self._format.decode(self._filename)
         except FileNotFoundError:
             raise StorageLookupError(
-                f"can not read {self._filename}", files=[self._filename]
+                f"cannot read {self._filename}", files=[self._filename]
             )
 
 

--- a/src/inline_snapshot/_new_adapter.py
+++ b/src/inline_snapshot/_new_adapter.py
@@ -151,7 +151,7 @@ class NewAdapter:
             return old_value
 
         if isinstance(new_value, CustomUnmanaged):
-            raise UsageError("unmanaged values can not be compared with snapshots")
+            raise UsageError("unmanaged values cannot be compared with snapshots")
 
         if (
             type(old_value) is type(new_value)

--- a/src/inline_snapshot/_snapshot/undecided_value.py
+++ b/src/inline_snapshot/_snapshot/undecided_value.py
@@ -74,7 +74,7 @@ class AstToCustom:
 
 class ValueToCustom:
     """this implementation is for cpython <= 3.10 only.
-    It works similar to AstToCustom but can not handle calls
+    It works similar to AstToCustom but cannot handle calls
     """
 
     context: AdapterContext

--- a/src/inline_snapshot/_snapshot_session.py
+++ b/src/inline_snapshot/_snapshot_session.py
@@ -329,7 +329,7 @@ class SnapshotSession:
             flags = cli_flags
             if parallel_run and flags - {"disable"}:
                 error(
-                    f"--inline-snapshot={','.join(flags)} can not be combined with xdist"
+                    f"--inline-snapshot={','.join(flags)} cannot be combined with xdist"
                 )
 
         state().flags = flags
@@ -346,10 +346,10 @@ class SnapshotSession:
 
         if "disable" in flags and flags != {"disable"}:
             error(
-                f"--inline-snapshot=disable can not be combined with other flags ({', '.join(flags-{'disable'})})"
+                f"--inline-snapshot=disable cannot be combined with other flags ({', '.join(flags-{'disable'})})"
             )
 
-        # disable inline-snapshot if it can not be used in the current context
+        # disable inline-snapshot if it cannot be used in the current context
         if parallel_run and "disable" not in flags:
             state().active = False
             state().disable_reason = "xdist"

--- a/src/inline_snapshot/plugin/_default_plugin.py
+++ b/src/inline_snapshot/plugin/_default_plugin.py
@@ -1,5 +1,6 @@
 import ast
 import datetime
+import sys
 from collections import Counter
 from collections import defaultdict
 from dataclasses import MISSING
@@ -13,6 +14,7 @@ from types import BuiltinFunctionType
 from types import FunctionType
 from typing import Any
 from typing import Dict
+from typing import NewType
 
 from inline_snapshot._customize._builder import Builder
 from inline_snapshot._customize._custom_code import ImportFrom
@@ -70,6 +72,23 @@ class InlineSnapshotPlugin:
             return builder.create_code(
                 qualname, imports=[ImportFrom(value.__module__, name)]
             )
+
+    if sys.version_info >= (3, 10):
+
+        @customize
+        def typing_newtype_handler(
+            self, value, builder: Builder, local_vars: Dict[str, Any]
+        ):
+            if type(value) is NewType:
+                for name, local_value in local_vars.items():
+                    if local_value is value:
+                        return builder.create_code(name)
+
+                qualname = value.__qualname__.split("[")[0]  # type: ignore[attr-defined]
+                name = qualname.split(".")[0]
+                return builder.create_code(
+                    qualname, imports=[ImportFrom(value.__module__, name)]
+                )
 
     @customize
     def builtin_function_handler(self, value, builder: Builder):
@@ -150,7 +169,7 @@ class InlineSnapshotPlugin:
             set_values = sorted(set_values)
             is_sorted = True
         except TypeError:
-            # can not be sorted by value
+            # cannot be sorted by value
             pass
 
         set_values = list(map(repr, set_values))
@@ -220,6 +239,10 @@ class InlineSnapshotPlugin:
             for field in fields(value):  # type: ignore
                 if field.repr:
                     field_value = getattr(value, field.name)
+
+                    if sys.version_info >= (3, 10):
+                        if isinstance(field.type, NewType):
+                            field_value = builder.create_call(field.type, [field_value])
 
                     if field.default != MISSING:
                         field_value = builder.with_default(field_value, field.default)
@@ -345,6 +368,12 @@ else:
                     if field.repr:
                         field_value = getattr(value, field.name)
 
+                        if sys.version_info >= (3, 10):
+                            if isinstance(field.type, NewType):
+                                field_value = builder.create_call(
+                                    field.type, [field_value]
+                                )
+
                         if field.default is not attrs.NOTHING:
 
                             default_value = (
@@ -401,17 +430,19 @@ else:
                     if getattr(field, "repr", True):
                         field_value = getattr(value, name)
 
-                        if (
-                            field.default is not PydanticUndefined
-                            and field.default == field_value
-                        ):
-                            field_value = builder.with_default(
-                                field_value, field.default
-                            )
+                        type_ = field.annotation
 
-                        elif field.default_factory is not None:
+                        if sys.version_info >= (3, 10):
+                            if isinstance(type_, NewType):
+                                field_value = builder.create_call(type_, [field_value])
+
+                        if field.default_factory is not None:
                             field_value = builder.with_default(
                                 field_value, field.default_factory()
+                            )
+                        elif field.default is not PydanticUndefined:
+                            field_value = builder.with_default(
+                                field_value, field.default
                             )
 
                         kwargs[name] = field_value

--- a/tests/adapter/test_call.py
+++ b/tests/adapter/test_call.py
@@ -109,7 +109,7 @@ def test_thing():
         raises=snapshot(
             """\
 TypeError:
-can not call CustomCode('BadCall')()\
+cannot call CustomCode('BadCall')()\
 """
         ),
     )

--- a/tests/adapter/test_dataclass.py
+++ b/tests/adapter/test_dataclass.py
@@ -1,3 +1,7 @@
+import sys
+
+import pytest
+
 from inline_snapshot import snapshot
 from inline_snapshot.extra import warns
 from inline_snapshot.testing._example import Example
@@ -772,3 +776,93 @@ def test_A():
             }
         ),
     )
+
+
+@pytest.mark.skipif(
+    sys.version_info < (3, 10),
+    reason="NewType is a function in 3.9 and cannot be checked with isinstance or serialized to code",
+)
+def test_dataclass_newtype():
+    Example(
+        """\
+from dataclasses import dataclass
+from typing import NewType
+from inline_snapshot import snapshot
+
+SomeID = NewType("SomeID", int)
+
+@dataclass
+class Something:
+    some_id: SomeID
+
+def test_something():
+    a = Something(some_id=SomeID(1))
+    assert a == snapshot()
+"""
+    ).run_inline(
+        ["--inline-snapshot=create"],
+        changed_files=snapshot(
+            {
+                "tests/test_something.py": """\
+from dataclasses import dataclass
+from typing import NewType
+from inline_snapshot import snapshot
+
+SomeID = NewType("SomeID", int)
+
+@dataclass
+class Something:
+    some_id: SomeID
+
+def test_something():
+    a = Something(some_id=SomeID(1))
+    assert a == snapshot(Something(some_id=SomeID(1)))
+"""
+            }
+        ),
+    ).run_inline()
+
+
+@pytest.mark.skipif(
+    sys.version_info < (3, 10),
+    reason="NewType is a function in 3.9 and cannot be checked with isinstance or serialized to code",
+)
+def test_attrs_newtype():
+    Example(
+        """\
+import attrs
+from typing import NewType
+from inline_snapshot import snapshot
+
+SomeID = NewType("SomeID", int)
+
+@attrs.define
+class Something:
+    some_id: SomeID
+
+def test_something():
+    a = Something(some_id=SomeID(1))
+    assert a == snapshot()
+"""
+    ).run_inline(
+        ["--inline-snapshot=create"],
+        changed_files=snapshot(
+            {
+                "tests/test_something.py": """\
+import attrs
+from typing import NewType
+from inline_snapshot import snapshot
+
+SomeID = NewType("SomeID", int)
+
+@attrs.define
+class Something:
+    some_id: SomeID
+
+def test_something():
+    a = Something(some_id=SomeID(1))
+    assert a == snapshot(Something(some_id=SomeID(1)))
+"""
+            }
+        ),
+    ).run_inline()

--- a/tests/adapter/test_general.py
+++ b/tests/adapter/test_general.py
@@ -63,7 +63,7 @@ def test_thing():
         raises=snapshot(
             """\
 UsageError:
-unmanaged values can not be compared with snapshots\
+unmanaged values cannot be compared with snapshots\
 """
         ),
     )

--- a/tests/external/test_external.py
+++ b/tests/external/test_external.py
@@ -732,7 +732,7 @@ def test_a():
         error=snapshot(
             """\
 >       assert "hi" == external(".json") == external(".txt")
-E       inline_snapshot._exceptions.UsageError: you can not compare external(...) with external(...)
+E       inline_snapshot._exceptions.UsageError: you cannot compare external(...) with external(...)
 """
         ),
         returncode=1,
@@ -752,7 +752,7 @@ def test_a():
         error=snapshot(
             """\
 >       assert "hi" == external(".json") == snapshot(".txt")
-E       inline_snapshot._exceptions.UsageError: you can not compare external(...) with snapshot(...)
+E       inline_snapshot._exceptions.UsageError: you cannot compare external(...) with snapshot(...)
 """
         ),
         returncode=1,
@@ -835,7 +835,7 @@ You can also use --inline-snapshot=review to approve the changes interactively\
             """\
 >       assert "hi" == external()
 >           raise UsageError(
-E           inline_snapshot._exceptions.UsageError: can not load external object from an non existing location 'uuid:'
+E           inline_snapshot._exceptions.UsageError: cannot load external object from a non-existing location 'uuid:'
 """
         ),
         report=snapshot(""),

--- a/tests/test_builder.py
+++ b/tests/test_builder.py
@@ -28,7 +28,7 @@ def test_a():
         raises=snapshot(
             """\
 UsageError:
-default value can not be an Custom value\
+default value cannot be a Custom value\
 """
         ),
     )

--- a/tests/test_dirty_equals.py
+++ b/tests/test_dirty_equals.py
@@ -1,10 +1,7 @@
-import pytest
-
 from inline_snapshot._inline_snapshot import snapshot
 from inline_snapshot.testing._example import Example
 
 
-@pytest.mark.xfail
 def test_dirty_equals_repr():
     Example(
         """\
@@ -20,13 +17,12 @@ def test_something():
         raises=snapshot(
             """\
 UsageError:
-inline-snapshot uses `copy.deepcopy` to copy objects,
-but the copied object is not equal to the original one:
+Customized value does not match original value:
 
-original: [HasRepr(IsStr, '< type(obj) can not be compared with == >')]
-copied:   [HasRepr(IsStr, '< type(obj) can not be compared with == >')]
+original_value=IsStr()
 
-Please fix the way your object is copied or your __eq__ implementation.
+customized_value=IsStr()
+customized_representation=CustomCall(function=CustomCode('IsStr'), args=[], kwargs={})
 """
         ),
     )

--- a/tests/test_pydantic.py
+++ b/tests/test_pydantic.py
@@ -1,3 +1,5 @@
+import sys
+
 import pydantic
 import pytest
 
@@ -23,7 +25,7 @@ def test_pydantic():
     assert m.dict()==snapshot()
 
     """
-    ).run_pytest(
+    ).run_inline(
         ["--inline-snapshot=create"],
         changed_files=snapshot(
             {
@@ -46,8 +48,7 @@ def test_pydantic():
 """
             }
         ),
-        returncode=1,
-    ).run_pytest(
+    ).run_inline(
         ["--inline-snapshot=disable"]
     )
 
@@ -66,7 +67,7 @@ class container(BaseModel):
 def test():
     assert container(a=1,b=5) == snapshot()
 """
-    ).run_pytest(
+    ).run_inline(
         ["--inline-snapshot=create"],
         changed_files=snapshot(
             {
@@ -83,8 +84,7 @@ def test():
 """
             }
         ),
-        returncode=1,
-    ).run_pytest(
+    ).run_inline(
         ["--inline-snapshot=disable"]
     )
 
@@ -104,7 +104,7 @@ class A(BaseModel):
 def test_something():
     assert A(a=1) == snapshot(A(a=1,b=2,c=[]))
 """
-    ).run_pytest(
+    ).run_inline(
         ["--inline-snapshot=update"],
         changed_files=snapshot(
             {
@@ -139,7 +139,7 @@ def test_something():
     for _ in [1,2]:
         assert A(a=1) == snapshot(A(a=1))
 """
-    ).run_pytest(
+    ).run_inline(
         changed_files=snapshot({}),
     )
 
@@ -161,7 +161,7 @@ def test_something():
     for a in [1,2]:
         assert A(a=2) == snapshot(A.from_str("1"))
 """
-    ).run_pytest(
+    ).run_inline(
         ["--inline-snapshot=fix"],
         changed_files=snapshot(
             {
@@ -182,13 +182,104 @@ def test_something():
 """
             }
         ),
-        returncode=1,
     )
 
 
 @pytest.mark.skipif(
+    sys.version_info < (3, 10),
+    reason="NewType is a function in 3.9 and cannot be checked with isinstance or serialized to code",
+)
+def test_pydantic_newtype():
+    Example(
+        """\
+from typing import NewType
+from pydantic import BaseModel
+from inline_snapshot import snapshot
+
+
+def test_something():
+    SomeID = NewType("SomeID", int)
+
+    class Something(BaseModel):
+        some_id: SomeID
+    a = Something(some_id=SomeID(1))
+    assert a == snapshot()
+"""
+    ).run_inline(
+        ["--inline-snapshot=create"],
+        changed_files=snapshot(
+            {
+                "tests/test_something.py": """\
+from typing import NewType
+from pydantic import BaseModel
+from inline_snapshot import snapshot
+
+
+def test_something():
+    SomeID = NewType("SomeID", int)
+
+    class Something(BaseModel):
+        some_id: SomeID
+    a = Something(some_id=SomeID(1))
+    assert a == snapshot(Something(some_id=SomeID(1)))
+"""
+            }
+        ),
+    ).run_inline()
+
+
+@pytest.mark.skipif(
+    sys.version_info < (3, 10),
+    reason="NewType is a function in 3.9 and cannot be checked with isinstance or serialized to code",
+)
+def test_pydantic_newtype_import():
+    Example(
+        {
+            "some_id.py": """\
+from typing import NewType
+from pydantic import BaseModel
+from inline_snapshot import snapshot
+
+SomeID = NewType("SomeID", int)
+
+class Something(BaseModel):
+    some_id1: SomeID
+    some_id2: SomeID=SomeID(5)
+
+
+def create():
+    return Something(some_id1=SomeID(1),some_id2=SomeID(5))
+                """,
+            "test_something.py": """\
+from inline_snapshot import snapshot
+from some_id import create
+
+def test_something():
+    assert create() == snapshot()
+""",
+        }
+    ).run_inline(
+        ["--inline-snapshot=create"],
+        changed_files=snapshot(
+            {
+                "test_something.py": """\
+from inline_snapshot import snapshot
+from some_id import create
+
+from some_id import SomeID
+from some_id import Something
+
+def test_something():
+    assert create() == snapshot(Something(some_id1=SomeID(1)))
+"""
+            }
+        ),
+    ).run_inline()
+
+
+@pytest.mark.skipif(
     pydantic.version.VERSION.startswith("1."),
-    reason="pydantic 1 can not compare C[int]() with C()",
+    reason="pydantic 1 cannot compare C[int]() with C()",
 )
 def test_pydantic_generic_class():
     Example(

--- a/tests/test_pytest_diff_fix.py
+++ b/tests/test_pytest_diff_fix.py
@@ -7,7 +7,7 @@ from inline_snapshot.testing._example import Example
 
 @pytest.mark.skipif(
     not is_pytest_compatible(),
-    reason="pytest assert rewriting and report can not be used at the same time",
+    reason="pytest assert rewriting and report cannot be used at the same time",
 )
 def test_pytest_diff_fix():
 

--- a/tests/test_pytest_plugin.py
+++ b/tests/test_pytest_plugin.py
@@ -286,12 +286,12 @@ def test_a():
 
     result = project.run("--inline-snapshot=disable,fix")
     assert result.stderr.str().strip() == snapshot(
-        "ERROR: --inline-snapshot=disable can not be combined with other flags (fix)"
+        "ERROR: --inline-snapshot=disable cannot be combined with other flags (fix)"
     )
 
     result = project.run("--inline-snapshot=disable,review")
     assert result.stderr.str().strip() == snapshot(
-        "ERROR: --inline-snapshot=disable can not be combined with other flags (review)"
+        "ERROR: --inline-snapshot=disable cannot be combined with other flags (review)"
     )
 
 

--- a/tests/test_xdist.py
+++ b/tests/test_xdist.py
@@ -15,7 +15,7 @@ def test_a():
     result = project.run("--inline-snapshot=create", "-n=auto")
 
     assert "\n".join(result.stderr.lines).strip() == snapshot(
-        "ERROR: --inline-snapshot=create can not be combined with xdist"
+        "ERROR: --inline-snapshot=create cannot be combined with xdist"
     )
 
     assert result.ret == 4
@@ -66,7 +66,7 @@ def test_a():
     assert result.report == snapshot("")
 
     assert result.stderr.lines == snapshot(
-        ["ERROR: --inline-snapshot=fix can not be combined with xdist", ""]
+        ["ERROR: --inline-snapshot=fix cannot be combined with xdist", ""]
     )
 
     assert result.ret == 4

